subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java

```diff
@@ -211,34 +212,54 @@ public void visitProducerTasks(Action<? super Task> visitor) {
          * and instead always executes against the given {@code value}.
          */
         static <T, A> SideEffect<T> fixed(A value, SideEffect<A> sideEffect) {
-            return new FixedSideEffect<>(value, sideEffect);
+            return EmptySideEffect.isEmpty(sideEffect) ? EmptySideEffect.value() : new FixedSideEffect<>(value, sideEffect);
+        }
+
+        static <T, A> SideEffect<T> fixedFrom(Value<A> value) {
+            if (value.isMissing()) {
+                return EmptySideEffect.value();
+            }
+
+            SideEffect<? super A> sideEffect = value.getSideEffect();
+            return EmptySideEffect.isEmpty(sideEffect) ? EmptySideEffect.value() : fixed(value.getWithoutSideEffect(), sideEffect);
         }
 
         /**
          * Creates a new side effect that executes given side effects sequentially.
          */
-        static <T> SideEffect<T> composite(Iterable<SideEffect<? super T>> sideEffects) {
-            // TODO: optimize for empty and single-value cases
-            return new CompositeSideEffect<>(sideEffects);
+        static <T> SideEffect<T> composite(Iterable<SideEffect<T>> sideEffects) {
+            return CompositeSideEffect.of(sideEffects);
```

Comment:
I think we should handle the case here when the list of side effects only has one element, and then return this one element from the factory methods.

Reply:
We already do this in the `CompositeSideEffect.of`. It optimizes for the case of 0 or 1 elements.